### ISSUE TITLE
Numba, in nopython mode, cannot inline @vectorize functions

### ISSUE THREAD

**Issue by FedericoV** [numba/numba#1300]
Quick example.  I'm using the latest numba (0.20).

I defined my own gaussian logpdf because the SciPy one was too slow.

```
@numba.vectorize(["f8(f8, f8, f8)"], nopython=True)
def numba_lognormalpdf1(x, mean, sigma):
    return -0.5*np.log(2*np.pi*sigma**2)- (x-mean)**2/sigma**2/2.0

@numba.jit(nopython=True)
def test_numba1(x, mean, scale, y):
    b = numba_lognormalpdf1(x, mean, scale)
    return b + y

@numba.jit(nopython=True)
def numba_lognormalpdf2(x, mean, sigma):
    return -0.5*np.log(2*np.pi*sigma**2)- (x-mean)**2/sigma**2/2.0

@numba.jit(nopython=True)
def test_numba2(x, mean, scale, y):
    b = numba_lognormalpdf2(x, mean, scale)
    return b + y
```

If I then do:

```
test_numba1(0.0, 0.0, 1.0, 2.0)
---------------------------------------------------------------------------
TypingError                               Traceback (most recent call last)
<ipython-input-23-87b4fe59b97e> in <module>()
----> 1 test_numba1(0.0, 0.0, 1.0, 2.0)

/home/federico/anaconda/lib/python2.7/site-packages/numba/dispatcher.pyc in _compile_for_args(self, *args, **kws)
    157         assert not kws
    158         sig = tuple([self.typeof_pyval(a) for a in args])
--> 159         return self.compile(sig)
    160 
    161     def inspect_llvm(self, signature=None):

/home/federico/anaconda/lib/python2.7/site-packages/numba/dispatcher.pyc in compile(self, sig)
    320                                           self.py_func,
    321                                           args=args, return_type=return_type,
--> 322                                           flags=flags, locals=self.locals)
    323 
    324             # Check typing error if object mode is used

/home/federico/anaconda/lib/python2.7/site-packages/numba/compiler.pyc in compile_extra(typingctx, targetctx, func, args, return_type, flags, locals, library)
    592     pipeline = Pipeline(typingctx, targetctx, library,
    593                         args, return_type, flags, locals)
--> 594     return pipeline.compile_extra(func)
    595 
    596 

/home/federico/anaconda/lib/python2.7/site-packages/numba/compiler.pyc in compile_extra(self, func)
    315                 raise e
    316 
--> 317         return self.compile_bytecode(bc, func_attr=self.func_attr)
    318 
    319     def compile_bytecode(self, bc, lifted=(), lifted_from=None,

/home/federico/anaconda/lib/python2.7/site-packages/numba/compiler.pyc in compile_bytecode(self, bc, lifted, lifted_from, func_attr)
    324         self.lifted_from = lifted_from
    325         self.func_attr = func_attr
--> 326         return self._compile_bytecode()
    327 
    328     def compile_internal(self, bc, func_attr=DEFAULT_FUNCTION_ATTRIBUTES):

/home/federico/anaconda/lib/python2.7/site-packages/numba/compiler.pyc in _compile_bytecode(self)
    579 
    580         pm.finalize()
--> 581         return pm.run(self.status)
    582 
    583 

/home/federico/anaconda/lib/python2.7/site-packages/numba/compiler.pyc in run(self, status)
    207                     # No more fallback pipelines?
    208                     if is_final_pipeline:
--> 209                         raise patched_exception
    210                     # Go to next fallback pipeline
    211                     else:

TypingError: Caused By:
Traceback (most recent call last):
  File "/home/federico/anaconda/lib/python2.7/site-packages/numba/compiler.py", line 201, in run
    res = stage()
  File "/home/federico/anaconda/lib/python2.7/site-packages/numba/compiler.py", line 414, in stage_nopython_frontend
    self.locals)
  File "/home/federico/anaconda/lib/python2.7/site-packages/numba/compiler.py", line 708, in type_inference_stage
    infer.build_constrain()
  File "/home/federico/anaconda/lib/python2.7/site-packages/numba/typeinfer.py", line 386, in build_constrain
    self.constrain_statement(inst)
  File "/home/federico/anaconda/lib/python2.7/site-packages/numba/typeinfer.py", line 506, in constrain_statement
    self.typeof_assign(inst)
  File "/home/federico/anaconda/lib/python2.7/site-packages/numba/typeinfer.py", line 538, in typeof_assign
    self.typeof_global(inst, inst.target, value)
  File "/home/federico/anaconda/lib/python2.7/site-packages/numba/typeinfer.py", line 603, in typeof_global
    loc=inst.loc)
TypingError: Untyped global name 'numba_lognormalpdf1'
File "<ipython-input-21-ae6bd434772e>", line 7

Failed at nopython (nopython frontend)
Untyped global name 'numba_lognormalpdf1'
File "<ipython-input-21-ae6bd434772e>", line 7
```

While:
```
In [24]:

test_numba2(0.0, 0.0, 1.0, 2.0)
Out[24]:
1.0810614667953273
```

**Comment by sklam** [comment]
When function signatures are provided in @numba.vectorize,  the resulting function is a numpy ufunc and cannot be used in nopython mode.  The newly introduced dufunc is only enabled if no signature is provided.  The following works:

```python


@numba.vectorize(nopython=True)
def numba_lognormalpdf1(x, mean, sigma):
    return -0.5*np.log(2*np.pi*sigma**2)- (x-mean)**2/sigma**2/2.0


@numba.jit(nopython=True)
def test_numba1(x, mean, scale, y):
    b = numba_lognormalpdf1(x, mean, scale)
    return b + y


print(test_numba1(0.0, 0.0, 1.0, 2.0))
```

Also, notice the difference in:

```python

@numba.vectorize(nopython=True)
def numba_lognormalpdf_dufunc(x, mean, sigma):
    return -0.5*np.log(2*np.pi*sigma**2)- (x-mean)**2/sigma**2/2.0

@numba.vectorize(["f8(f8, f8, f8)"], nopython=True)
def numba_lognormalpdf_numpy(x, mean, sigma):
    return -0.5*np.log(2*np.pi*sigma**2)- (x-mean)**2/sigma**2/2.0

print(numba_lognormalpdf_dufunc)
print(numba_lognormalpdf_numpy)
```
output:
```
<numba._DUFunc 'numba_lognormalpdf_dufunc'>
<ufunc 'numba_lognormalpdf_numpy'>
```

**Comment by seibert** [comment]
FYI, this is a limitation ultimately we would like to remove.  :)

**Comment by FedericoV** [comment]
Aha, thanks a lot for the heads up.  Usually, if I know the types of the functions I type them all ahead of time just to have some type checking as an extra compile check - but I guess in this case there's absolutely no advantage to do so.

Should I close this?

**Comment by pitrou** [comment]
This will be fixed by PR #1615.
